roborock/version_1_apis/roborock_local_client_v1.py

```diff
@@ -121,20 +110,56 @@ async def async_disconnect(self) -> None:
         async with self._mutex:
             self._sync_disconnect()
 
-    async def hello(self):
+    def _set_encoder_decoder(self):
+        """Updates the encoder decoder. For L01 these are updated with nonces after the first hello."""
+        self._encoder = create_local_encoder(self.device_info.device.local_key, self._connect_nonce, self._ack_nonce)
+        self._decoder = create_local_decoder(self.device_info.device.local_key, self._connect_nonce, self._ack_nonce)
+
+    async def _do_hello(self, version: str) -> bool:
+        """Perform the initial handshaking."""
+        self._logger.debug(f"Attempting to use the {version} protocol for client {self.device_info.device.duid}...")
+        request = RoborockMessage(
+            protocol=RoborockMessageProtocol.HELLO_REQUEST,
+            version=version.encode(),
+            random=self._connect_nonce,
+            seq=1,
+        )
         try:
-            return await self._send_message(
-                roborock_message=_HELLO_REQUEST_MESSAGE,
-                request_id=_HELLO_REQUEST_MESSAGE.seq,
+            response = await self._send_message(
+                roborock_message=request,
+                request_id=request.seq,
                 response_protocol=RoborockMessageProtocol.HELLO_RESPONSE,
             )
-        except Exception as e:
-            self._logger.error(e)
+            self._ack_nonce = response.random
+            self._set_encoder_decoder()
+            self._version = version
+            self._logger.debug(f"Client {self.device_info.device.duid} speaks the {version} protocol.")
+            return True
+        except RoborockException as e:
+            self._logger.debug(
+                f"Client {self.device_info.device.duid} did not respond or does not speak the {version} protocol. {e}"
+            )
+            return False
+
+    async def hello(self):
+        """Send hello to the device to negotiate protocol."""
+        if self._version:
+            # version is forced
+            if not await self._do_hello(self._version):
+                raise RoborockException(f"Failed to connect to device with protocol {self._version}")
+        else:
+            # try 1.0, then L01
+            if not await self._do_hello("1.0"):
+                if not await self._do_hello("L01"):
+                    raise RoborockException("Failed to connect to device with any known protocol")
 
     async def ping(self) -> None:
+        # Realistically, this should be set here, but this is to be safe and for typing.
+        version = b"1.0" if self._version is None else self._version.encode()
```

Comment:
How about moving this to a property shared between the two callers of send_message? The logic is inversed in each so hard to tell if its different on purpose, but i think its basically the same thing just with an arbitrary ordering.